solidity/contracts/MailboxV2.sol

```diff
@@ -0,0 +1,225 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+
+// ============ Internal Imports ============
+import {Versioned} from "./upgrade/Versioned.sol";
+import {MerkleLib} from "./libs/Merkle.sol";
+import {MessageV2} from "./libs/MessageV2.sol";
+import {TypeCasts} from "./libs/TypeCasts.sol";
+import {IMessageRecipient} from "../interfaces/IMessageRecipient.sol";
+import {IInterchainSecurityModule, ISpecifiesInterchainSecurityModule} from "../interfaces/IInterchainSecurityModule.sol";
+import {IMailboxV2} from "../interfaces/IMailboxV2.sol";
+
+// ============ External Imports ============
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import {ReentrancyGuardUpgradeable} from "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+
+contract MailboxV2 is
+    IMailboxV2,
+    OwnableUpgradeable,
+    ReentrancyGuardUpgradeable,
+    Versioned
+{
+    // ============ Libraries ============
+
+    using MerkleLib for MerkleLib.Tree;
+    using MessageV2 for bytes;
+    using TypeCasts for bytes32;
+    using TypeCasts for address;
+
+    // ============ Constants ============
+
+    // Maximum bytes per message = 2 KiB (somewhat arbitrarily set to begin)
+    uint256 public constant MAX_MESSAGE_BODY_BYTES = 2 * 2**10;
+    // Domain of chain on which the contract is deployed
+    uint32 public immutable localDomain;
+
+    // ============ Public Storage ============
+
+    // The default ISM, used if the recipient fails to specify one.
+    IInterchainSecurityModule public defaultModule;
+    // An incremental merkle tree used to store outbound message IDs.
+    MerkleLib.Tree public tree;
+    // Mapping of message ID to whether or not that message has been delivered.
+    mapping(bytes32 => bool) public delivered;
+
+    // ============ Upgrade Gap ============
+
+    // gap for upgrade safety
+    uint256[47] private __GAP;
+
+    // ============ Events ============
+
+    /**
+     * @notice Emitted when the default ISM is updated
+     * @param module The new default ISM
+     */
+    event DefaultModuleSet(address indexed module);
+
+    /**
+     * @notice Emitted when a new message is dispatched via Hyperlane
+     * @param messageId The unique message identifier
+     * @param message Raw bytes of message
+     */
+    event Dispatch(bytes32 indexed messageId, bytes message);
+
+    /**
+     * @notice Emitted when a Hyperlane message is delivered
+     * @param messageId The unique message identifier
+     */
+    event Process(bytes32 indexed messageId);
+
+    // ============ Constructor ============
+
+    // solhint-disable-next-line no-empty-blocks
+    constructor(uint32 _localDomain) {
```

Comment:
Do make mailboxes have the same address on every chain I think we need to move this to the initliazer?

Reply:
I guess maybe having mailboxes have the same address is untenable anyways in a permissionless deployment, but still seems possibly desirable?